proto/p4/p4runtime.md

```diff
@@ -0,0 +1,294 @@
+pseudocode as an example:
+
+```python
+intended_value = value
+
+status = server.write(intended_value, p4_object)
+observed_value = server.read(p4_object)
+
+assert(intended_value == observed_value)
+```
+
+To ensure read-write symmetry, the rest of the doc tries to offer canonical
+representations for various data types, but this principle should be thought of
+where it falls short. Ensuring this will allow a controller software to recover
+programmatically from failures that can affect the switch stack software,
+communication channel, or the controller replicas. If Read RPC returns a
+semantically-same but syntactically-different response then the client would
+have to canonicalize the read values to check its internal state, which only
+pushes the protocol's complexities to the client implementations.
+
+### Map-like message sequences
+
+Map-like messages, i.e., messages that contain an ID and a value (e.g.,
+`field_id`, and a field value in `FieldMatch`), are common in P4Runtime. Because
+every entry contains a unique ID, reordering them yields semantically-equivalent
+messages.
+
+For map-like data structures, we make an exception to the "syntactically-same"
+rule and use semantic equivalence to avoid requiring the switch to remember the
+field order. Therefore, the controller is expected to use a check similar to the
+following Python snippet in deciding the equality of two map-like structures:
+
+```python
+def Equal(map_like_x, map_like_y):
+  # Parse structures into maps.
+  map_x = dict()
+  for i, v := map_like_x:
+    map_x[i] = v
+
+  map_y = dict()
+  for i, v := map_like_x:
+    map_y[i] = v
+
+  # Check map equivalence.
+  return map_x == map_y
+```
+
+## Byte strings
+
+P4Runtime data sizes may be large enough to not fit in common data types that
+largely follow a host machines word sizes, i.e., 32 or 64 bits. The P4 language
+does not put any limit on the header fields or action parameters a switch can
+operate on, and it is up to a P4 program to determine them. Due to the
+flexibility of data sizes, P4Runtime declares them to be byte strings of any
+integer size.
+
+### Byte string length
+
+The server ensures that the byte string for each data type has the correct size
+with respect to the definition of the data type within the P4 program, and
+returns *error* otherwise:
+
+```python
+bytestring_length = (num_bits_in_bitfield_in_p4_program + 7) / 8
+```
+
+### Byte order
+
+For all byte strings with size larger than 0 P4Runtime uses big endian (i.e.,
+network) byte order.
+
+## Value formats and constraints for P4Runtime fields
+
+This section describes some requirements on the values the fields in the
+P4Runtime can have. This is to ensure a canonical representation as described
+earlier. The server should handle fields as follows (using p4runtime.proto
+format):
+
+### `TableEntry.match`, `FieldMatch.field_id`, and other fields of the
+`FieldMatch`
```

Comment:
I've just noticed this. You cannot split a header across 2 lines like this or the formatting is wrong with the Github markdown rendered. Would be great if you could amend your commit to remove the line break.